proto/cosmos/base/tendermint/v1beta1/query.proto

```diff
@@ -136,3 +141,47 @@ message Module {
   // checksum
   string sum = 3;
 }
+
+// ABCIQueryRequest defines the request structure for the ABCIQuery gRPC query.
+message ABCIQueryRequest {
+  bytes  data   = 1;
+  string path   = 2;
```

Comment:
How does this handle recursion? e.g. 
-  here set path = `/cosmos.base.terndermint.v1beta.Query/ABCIQuery`?
- or directly hit the tm rpc endpoint abci_query with path = `/cosmos.base.terndermint.v1beta.Query/ABCIQuery`?

Reply:
I'm not sure this is a non-issue. Your test has `expectedCode: 0,`, but the test name says "invalid", and I think I would also expect an error here.

Also, there's a gRPC broadcast endpoint. What happens if you put that endpoint as the `path`? Can we maybe add a test case:

```go
			req: &tmservice.ABCIQueryRequest{
				Path: "/cosmos.tx.v1beta1.Service/BroadcastTx",
				Data: <an actual well-encoded and correct tx>,
			},
```

The desired output is that it errors, but I'm not sure in practice how this turns out.

Reply:
Ahhh I didn't realize how this could be recursive because of the initial gRPC call! Ok this concern makes total sense.

Inspecting the recursive query with path `/cosmos.base.tendermint.v1beta1.Service/ABCIQuery`, I notice it gets executed 4 times:

```
QUERY PATH /cosmos.base.tendermint.v1beta1.Service/ABCIQuery
QUERY PATH /cosmos.base.tendermint.v1beta1.Service/ABCIQuery
QUERY PATH /cosmos.base.tendermint.v1beta1.Service/ABCIQuery
QUERY PATH
```

Notice, the last path is empty and I suppose this is why the code is 0 and non-error. Idk why or how it gets executed only 3 times before becoming empty or even why it becomes empty to begin with. 

How do you feel about adding an explicit path check to the gRPC handler -- only allow valid paths: `app`, `store`, `p2p`, and `custom`. This should cover all cases.

Reply:
See latest commit